libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts

```diff
@@ -0,0 +1,384 @@
+    const id = getUniqueId()
+
+    const rpcRequest: JsonRpcRequestMessage = {
+      jsonrpc: JSON_RPC_VERSION,
+      id,
+      method,
+      params,
+      // ...(typeof params === 'undefined' ? null : { params }),
+    }
+
+    const promise = new Promise<JsonRpcSucessfulResponseMessage<TResult> | JsonRpcErrorResponseMessage<TErrorData>>(
+      (resolve, reject) => (this.completers[id] = { resolve, reject })
+    )
+
+    // Send the JSON RPC to the event source.
+    postMessageToWindow(this.eventTarget, WidgetMethodsEmit.PROVIDER_RPC_REQUEST, {
+      rpcRequest,
+    })
+
+    // Delete the completer within the timeout and reject the promise.
+    setTimeout(() => {
+      if (this.completers[id]) {
+        this.completers[id].reject(new Error(`RPC ID "${id}" timed out after ${this.timeoutMilliseconds} milliseconds`))
+        delete this.completers[id]
+      }
+    }, this.timeoutMilliseconds)
+
+    return promise
+  }
+
+  /**
+   * Send the JSON RPC and return the result.
+   * @param method method to send to the parent provider
+   * @param params parameters to send
+   */
+  public async send<TResult = any>(method: string, params: unknown[]): Promise<TResult> {
+    const response = await this.execute<TResult, any>(method, params)
+
+    if ('error' in response) {
+      throw new RpcError(response.error.code, response.error.message)
+    } else {
+      return response.result
+    }
+  }
+
+  /**
+   * Request the parent window to enable access to the user's web3 provider. Return accounts list immediately if already enabled.
+   */
+  public async enable(): Promise<string[]> {
+    if (this.enabled === null) {
+      const promise = (this.enabled = this.send('enable', []).catch((error) => {
+        // Clear this.enabled if it's this promise so we try again next call.
+        // this.enabled might be set from elsewhere if, e.g. the accounts changed event is emitted
+        if (this.enabled === promise) {
+          this.enabled = null
+        }
+        // Rethrow the error.
+        throw error
+      }))
+    }
+
+    return this.enabled
+  }
+
+  /**
+   * Backwards compatibility method for web3.
+   * @param payload payload to send to the provider
+   * @param callback callback to be called when the provider resolves
+   */
+  public async sendAsync(
+    payload: { method: string; params: unknown[] },
+    callback: (error: string | null, result: { method: string; params?: any[]; result: any } | any) => void
+  ): Promise<void> {
+    try {
+      const result = await this.execute(payload.method, payload.params)
+
+      callback(null, result)
+    } catch (error) {
+      callback(error, null)
+    }
+  }
+
+  /**
+   * Handle a Rpc Request
+   */
+  private handleRpcRequests = ({ rpcResponse }: ProviderRpcResponsePayload) => {
+    if (rpcResponse.id === undefined || rpcResponse.id === null) {
+      return
+    }
+
+    const completer = this.completers['' + rpcResponse.id]
```

Comment:
Why add it to an empty string?
Shorthand for converting the id itself to string?
I suggest less black magic and call id.toString() or String(id)

Reply:
not my code. I think is how they converted to string the number...

I prefer to touch the minimum of this library